File: gradle/mixed-java-kotlin-hello-world/src/main/java/demo/HelloWorld.java
package demo;

public class HelloWorld {
    public static void main(String[] args) {
        final KotlinGreetingJoiner example = new KotlinGreetingJoiner(new Greeter("Hi"));

        example.addName("Harry");
        example.addName("Ron");
        example.addName(null);
        example.addName("Hermione");

        System.out.println(example.getJoinedGreeting());
    }
}

File: gradle/mixed-java-kotlin-hello-world/src/main/kotlin/helloWorld.kt
package demo

import java.util.*

class KotlinGreetingJoiner(val greeter: Greeter) {

    val names = ArrayList<String?>()

    fun addName(name: String?) {
        names.add(name)
    }

    fun getJoinedGreeting(): String {
        return "${greeter.getGreeting()} ${names.filterNotNull().joinToString(separator = " and ")}"
    }
}
